packages/beeai-sdk/src/beeai_sdk/providers/agent.py

```diff
@@ -13,36 +13,221 @@
-    server.settings.host = os.getenv("HOST", "127.0.0.1")
-    server.settings.port = int(os.getenv("PORT", "8000"))
-    trace.set_tracer_provider(
-        tracer_provider=TracerProvider(
-            resource=Resource(attributes={SERVICE_NAME: server.name, SERVICE_NAMESPACE: "beeai-agent-provider"}),
-            active_span_processor=BatchSpanProcessor(SilentOTLPSpanExporter()),
-        )
-    )
-    with trace.get_tracer("beeai-sdk").start_as_current_span("agent-provider"):
+class Server:
+    _agent: Agent
+    _manifest: dict[str, Any]
+
+    def __init__(self, name: str | None = None):
+        self.server = ACPServer(name or "beeai")
+        self.decorated = False
+
+    def __call__(self):
         try:
-            await server.run_sse_async(timeout_graceful_shutdown=5)
+            asyncio.run(self.run_agent_provider())
         except KeyboardInterrupt:
-            pass
+            return
+        except Exception as e:
+            logger.error(f"Error occured {e}")
+
+    def agent(self) -> Callable:
+        """Decorator to register an agent."""
+        if self.decorated:
+            raise Error("Agent decorator already used")
+        self.decorated = True
+
+        def decorator(func: Callable) -> Callable:
+            signature = inspect.signature(func)
+            parameters = list(signature.parameters.values())
+
+            if len(parameters) == 0:
+                raise TypeError("The agent function must have at least 'input' argument")
+            if len(parameters) > 2:
+                raise TypeError("The agent function must have only 'input' and 'ctx' arguments")
+            if len(parameters) == 2 and parameters[1].name != "ctx":
+                raise TypeError("The second argument of the agent function must be 'ctx'")
+
+            input = parameters[0].annotation
+            output = signature.return_annotation
+
+            def create_agent_name_from_path():
+                """Create an agent name from the current path"""
+                cwd = os.getcwd()
+                hash_object = hashlib.md5(cwd.encode())
+                return hash_object.hexdigest()
+
+            self._manifest = self.read_agent_manifest()
+
+            name = self._manifest.get("name") or create_agent_name_from_path()
+
+            @functools.wraps(func)
+            async def sync_fn(*args, **kwargs):
+                ctx = syncify_object_dynamic(kwargs["ctx"])
+                return await anyio.to_thread.run_sync(partial(func, ctx=ctx) if len(parameters) == 2 else func, *args)
+
+            @functools.wraps(func)
+            async def async_fn(*args, **kwargs):
+                return await func(*args, **kwargs) if len(parameters) == 2 else await func(*args)
+
+            self._agent = Agent(
+                name=name,
+                description=self._manifest.get("description"),
+                input=input,
+                output=output if output is not inspect.Signature.empty else Output,
```

Comment:
👍 
Let's add a warning, the user may unintentionally end up with a schema too broad if they forget to specify the return type annotation.